src/crimson/common/chained_dispatchers.cc

```diff
@@ -0,0 +1,39 @@
+#include "crimson/common/chained_dispatchers.h"
+#include "crimson/net/Connection.h"
+#include "msg/Message.h"
+
+seastar::future<>
+ChainedDispatchers::ms_dispatch(crimson::net::Connection* conn,
+                                MessageRef m) {
+  return seastar::do_for_each(dispatchers, [conn, m](Dispatcher& dispatcher) {
+    return dispatcher.ms_dispatch(conn, m);
+  });
+}
+
+void
+ChainedDispatchers::ms_handle_accept(crimson::net::ConnectionRef conn) {
+  (void)seastar::do_for_each(dispatchers, [conn](Dispatcher& dispatcher) {
```

Comment:
You mean dispatchers may not necessarily totally independent of each other, right?

Reply:
I mean dispatchers should not chain their futures with each other when handling accept/connect/reset/remote_reset events.

Reply:
I'm not sure why it's off topic -- the previous implementation ensured that all handlers would complete prior to dealing with another message.  This is actually incredibly important to ensure downstream ordering.  Otherwise, a stream of messages from the primary -> replica could be processed out of order and there isn't really anything the downstream dispatcher could do about it.  Peering and IO replication both rely on this.

Reply:
I think these need to return future<>'s which block on the dispatchers.

Reply:
@athanatos I excluded `ms_dispatch()` because I agree with your concerns about the message processing order.

I think it is a different issue regarding to the order across different kinds of simple events (accept/connect/reset/remote_reset) within the same dispatcher. A simplest solution with minimal performance overhead is to deliver them synchronously, which implies they should not be delayed by another chained dispatcher.

Of course we can have an asynchronous event delivery system which preserves order correctly with extra efforts, if necessary. And my understanding is that async-messenger implements both.